includes/class-wp-job-manager-post-types.php

```diff
@@ -659,6 +660,69 @@ public function job_feed() {
 		remove_filter( 'posts_search', 'get_job_listings_keyword_search' );
 	}
 
+	/**
+	 * Get filled jobs.
+	 *
+	 * @return array
+	 */
+	public function get_filled_job_listings(): array {
+		if ( ! get_option( 'job_manager_hide_filled_positions' ) ) {
+			return [];
+		}
+
+		$filled_jobs_transient = get_transient( 'hide_filled_jobs_transient' );
+		if ( false === $filled_jobs_transient ) {
+			$filled_jobs_transient = get_posts(
+				[
+					'post_type'      => 'job_listing',
+					'fields'         => 'ids',
+					'posts_per_page' => -1,
+					'meta_query'     => [
+						[
+							'key'     => '_filled',
+							'value'   => '1',
+							'compare' => '=',
+						],
+					],
+				]
+			);
+			set_transient( 'hide_filled_jobs_transient', $filled_jobs_transient, DAY_IN_SECONDS );
+		}
+		return $filled_jobs_transient;
+	}
+
+	/**
+	 * Maybe exclude expired and/or filled job listings from search and archive pages.
+	 *
+	 * @param $query WP_Query $query
+	 *
+	 * @return void
+	 */
+	public function maybe_hide_filled_expired_job_listings_from_search( WP_Query $query ): void {
+		$hide_filled_positions = get_option( 'job_manager_hide_filled_positions' );
+		$hide_expired          = get_option( 'job_manager_hide_expired' );
+
+		if ( ! $hide_filled_positions && ! $hide_expired ) {
+			return;
+		}
+
+		if (
+			! is_admin()
+			&& $query->is_main_query()
+			&& $query->is_search()
+			|| $query->is_archive() && 'job_listing' === $query->get( 'post_type' )
```

Comment:
I think this needs some grouping for the logic to be right:

```suggestion
			! is_admin()
			&& $query->is_main_query()
			&& ( $query->is_search() || $query->is_archive() ) 
			&& 'job_listing' === $query->get( 'post_type' )
```

Reply:
This breaks the method on WP search results pages, as the `post_type` returned is `any` rather than `job_listing` (i.e., we aren't necessarily searching within WP Job Manager itself, but using WordPress' default search and excluding results from its query.

That said, when testing in Jobify, I see that `expired` and `filled` jobs are hidden from taxonomy pages, and archive pages, like `/jobs`, hide these items by default as well. I'm not sure that these additions are actually needed, on further testing -- were you seeing behaviour on your end where they were showing, but shouldn't have been?

Reply:
Makes sense, didn't think about a generic search. I did see expired listings in the jobs archive page without the extra post status filter.

I also still see them with the latest approach of setting `expired` to non-public on post status registration. Seems like Jobify has a lot of custom code that might interfere? It's possible to test with any theme by enabling the templates with: `add_theme_support( 'job-manager-templates' );`. And might need saving on the Permalinks page to flush the rules.

Reply:
Gotcha. I ended up reverting to the previous `expired` implementation as the new one isn't able to check conditionals like `is_archive()`. 

Committed https://github.com/Automattic/WP-Job-Manager/pull/2449/commits/1a7b7da26d47c33255a23ea5e95da70773f6454a